pkg/controller/direct/logging/logmetric_controller.go

```diff
@@ -0,0 +1,253 @@
+	created, err := createRequest.Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("logMetric %s creation failed: %w", logMetric.Name, err)
+	}
+
+	log.V(2).Info("created logMetric", "logMetric", created)
+
+	resourceID := created.Name
+	if err := unstructured.SetNestedField(u.Object, resourceID, "spec", "resourceID"); err != nil {
+		return fmt.Errorf("setting spec.resourceID: %w", err)
+	}
+
+	status := &krm.LoggingLogMetricStatus{}
+	if err := logMetricStatusToKRM(created, status); err != nil {
+		return err
+	}
+
+	return setStatus(u, status)
+}
+
+func logMetricStatusToKRM(in *api.LogMetric, out *krm.LoggingLogMetricStatus) error {
+	out.CreateTime = &in.CreateTime
+	out.UpdateTime = &in.UpdateTime
+	out.MetricDescriptor = reduceAPItoKRM_MetricDescriptorStatus(in.MetricDescriptor)
+
+	return nil
+}
+
+func (a *logMetricAdapter) Update(ctx context.Context, u *unstructured.Unstructured) error {
+	update := new(api.LogMetric)
+	*update = *a.actual
+
+	if ValueOf(a.desired.Spec.Description) != a.actual.Description {
+		update.Description = ValueOf(a.desired.Spec.Description)
+	}
+	if ValueOf(a.desired.Spec.Disabled) != a.actual.Disabled {
+		update.Disabled = ValueOf(a.desired.Spec.Disabled)
+	}
+	if a.desired.Spec.Filter != a.actual.Filter {
+		// todo acpana: revisit UX, err out if filter of desired is empty
+		if a.desired.Spec.Filter != "" {
+			update.Filter = a.desired.Spec.Filter
+		} else {
+			// filter is a REQUIRED field
+			update.Filter = a.actual.Filter
+		}
+	}
+	if !compareMetricDescriptors(a.desired.Spec.MetricDescriptor, a.actual.MetricDescriptor) {
+		update.MetricDescriptor = expandKCCtoAPIForMetricDescriptor(a.desired.Spec.MetricDescriptor)
+	}
+
+	if !reflect.DeepEqual(a.desired.Spec.LabelExtractors, a.actual.LabelExtractors) {
+		update.LabelExtractors = a.desired.Spec.LabelExtractors
+	}
+
+	if !compareBucketOptions(a.desired.Spec.BucketOptions, a.actual.BucketOptions) {
+		update.BucketOptions = expandKCCtoAPIForBucketOptions(a.desired.Spec.BucketOptions)
+	}
+
+	if ValueOf(a.desired.Spec.ValueExtractor) != a.actual.ValueExtractor {
+		update.ValueExtractor = ValueOf(a.desired.Spec.ValueExtractor)
+	}
+
+	// todo acpana: add support for bucket_name
+
+	// DANGER: this is an upsert; it will create the LogMetric if it doesn't exists
+	// but this behavior is consistent with the DCL backed behavior we provide for this resource.
+	// todo acpana: look for / switch to a better method and/or use etags etc
+	actualUpdate, err := a.logMetricClient.Update(a.fullyQualifiedName(), update).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("logMetric update failed: %w", err)
+	}
+
+	status := &krm.LoggingLogMetricStatus{}
+	if err := logMetricStatusToKRM(actualUpdate, status); err != nil {
+		return err
+	}
+
+	// actualUpdate may not contain the description for the metric descriptor.
+	if update.Description != "" {
+		status.MetricDescriptor.Description = &update.Description
+	}
```

Comment:
Highlighting this bc I did think it was a bit odd that the actualUpdate lacked the `Description` field.